File: week1_baseline/ruby/09_global_executable/lib/boukensha_loader.rb
# BoukenshaLoader resolves which step folder to load from, then boots the REPL.
#
# Two independent settings are resolved, each with its own precedence:
#
#   lib path (which *step* to load):
#     1. BOUKENSHA_PATH environment variable
#     2. `path=` line in ~/.boukensharc (or the whole file, for the legacy
#        bare-path format — see parse_rc)
#     3. The lib/ directory bundled inside this gem (step 8 — the latest release)
#
#   config directory (settings.yaml, .env, system.md):
#     1. BOUKENSHA_DIR environment variable
#     2. `dir=` line in ~/.boukensharc
#     3. ~/.boukensha (Config's own default — the loader does nothing, and
#        Config.new resolves this itself)
#
# ~/.boukensharc format:
#   Either a bare path on its own (legacy, treated as `path=`):
#     echo ~/Sites/boukensha/08_the_repl_loop > ~/.boukensharc
#   Or key=value lines, one setting per line:
#     path=~/Sites/boukensha/08_the_repl_loop
#     dir=~/projects/mybot/.boukensha
#
# Examples:
#   boukensha                                                              # uses bundled lib + ~/.boukensha
#   BOUKENSHA_PATH=~/Sites/boukensha/04_api_client boukensha              # loads step 4
#   BOUKENSHA_DIR=~/projects/mybot/.boukensha boukensha                   # custom config dir
#   echo ~/Sites/boukensha/08_the_repl_loop > ~/.boukensharc && boukensha # permanent step default
module BoukenshaLoader
  # Absolute path to this gem's own bundled boukensha lib.
  BUNDLED_LIB = File.expand_path("../boukensha.rb", __FILE__)

  # Reads ~/.boukensharc and returns a { "path" => ..., "dir" => ... } hash
  # (only the keys actually present). Supports two formats:
  #   - key=value lines (any subset of "path"/"dir", in any order)
  #   - a single bare path with no "=" anywhere in the file (legacy format,
  #     treated as if it were `path=<that value>`)
  def self.parse_rc
    rc = File.expand_path("~/.boukensharc")
    return {} unless File.exist?(rc)

    lines = File.read(rc).lines.map(&:strip).reject(&:empty?)
    return {} if lines.empty?

    if lines.none? { |line| line.include?("=") }
      return { "path" => lines.first }
    end

    lines.each_with_object({}) do |line, settings|
      key, value = line.split("=", 2)
      next unless key && value

      settings[key.strip] = value.strip
    end
  end

  def self.resolve
    rc = parse_rc

    # BOUKENSHA_DIR: env var wins, then ~/.boukensharc's `dir=`, then
    # Config's own default (nothing to do here in that case).
    if !ENV["BOUKENSHA_DIR"] && rc["dir"] && !rc["dir"].empty?
      ENV["BOUKENSHA_DIR"] = File.expand_path(rc["dir"])
    end

    # 1. Env var wins.
    if ENV["BOUKENSHA_PATH"]
      dir  = File.expand_path(ENV["BOUKENSHA_PATH"])
      main = File.join(dir, "lib", "boukensha.rb")
      return main if File.exist?(main)

      abort <<~MSG
        boukensha: BOUKENSHA_PATH is set but no lib/boukensha.rb found at:
               #{dir}
               Make sure BOUKENSHA_PATH points to a step folder, e.g.:
               BOUKENSHA_PATH=~/Sites/boukensha/07_the_repl_loop boukensha
      MSG
    end

    # 2. ~/.boukensharc's `path=` (or legacy bare-path format)
    if rc["path"] && !rc["path"].empty?
      dir  = rc["path"]
      main = File.join(File.expand_path(dir), "lib", "boukensha.rb")
      return main if File.exist?(main)

      abort <<~MSG
        boukensha: ~/.boukensharc points to #{dir}
               but no lib/boukensha.rb was found there.
               Update ~/.boukensharc or remove it to use the bundled default.
      MSG
    end

    # 3. Bundled default.
    BUNDLED_LIB
  end

  def self.load_and_start_repl
    main = resolve
    step_dir = File.dirname(File.dirname(main))

    if ENV["BOUKENSHA_DEBUG"]
      puts "[boukensha] loading from: #{step_dir}"
      puts "[boukensha] config dir: #{ENV["BOUKENSHA_DIR"]}" if ENV["BOUKENSHA_DIR"]
    end

    require main

    unless Boukensha.respond_to?(:repl)
      abort <<~MSG
        boukensha: the step at #{step_dir}
               does not support the interactive REPL (added in step 7).
               Run its examples directly, e.g.:
                 ruby #{step_dir}/examples/*.rb
               Or point BOUKENSHA_PATH at step 7 or later.
      MSG
    end

    Boukensha.repl
  end
end

File: week1_baseline/python/09_global_executable/src/boukensha/boukensha_loader.py
"""Boukensha loader that resolves which step version to load and starts the REPL."""

import os
import sys
from pathlib import Path


def resolve():
    """Resolve which step's boukensha lib to load.

    Resolution order:
    1. BOUKENSHA_PATH environment variable (selects which step lib to load)
    2. ~/.boukensharc file (persistent step selection)
    3. Bundled default (this step's lib)

    Returns:
        str: Path to boukensha.py file to load
    """
    # 1. Check BOUKENSHA_PATH env var
    if os.environ.get("BOUKENSHA_PATH"):
        custom_path = os.path.expanduser(os.environ["BOUKENSHA_PATH"])
        lib_path = os.path.join(custom_path, "src", "boukensha", "__init__.py")

        if os.path.exists(lib_path):
            return lib_path

        sys.stderr.write(
            f"boukensha: BOUKENSHA_PATH is set but no src/boukensha/__init__.py found at:\n"
            f"       {custom_path}\n"
            f"       Make sure BOUKENSHA_PATH points to a step folder, e.g.:\n"
            f"       BOUKENSHA_PATH=~/Sites/boukensha/08_the_repl_loop boukensha\n"
        )
        sys.exit(1)

    # 2. Check ~/.boukensharc
    rc_path = os.path.expanduser("~/.boukensharc")
    if os.path.exists(rc_path):
        with open(rc_path, "r") as f:
            step_path = f.read().strip()

        if step_path:
            step_path = os.path.expanduser(step_path)
            lib_path = os.path.join(step_path, "src", "boukensha", "__init__.py")

            if os.path.exists(lib_path):
                return lib_path

            sys.stderr.write(
                f"boukensha: ~/.boukensharc points to {step_path}\n"
                f"       but no src/boukensha/__init__.py was found there.\n"
                f"       Update ~/.boukensharc or remove it to use the bundled default.\n"
            )
            sys.exit(1)

    # 3. Bundled default (this step's lib)
    default_lib = os.path.join(
        os.path.dirname(__file__),
        "__init__.py"
    )
    return default_lib


def load_and_start_repl():
    """Load the resolved boukensha step and start the REPL."""
    lib_path = resolve()
    # lib_path is: /path/to/step/src/boukensha/__init__.py
    # We need: step_dir = /path/to/step, src_dir = /path/to/step/src
    src_dir = os.path.dirname(os.path.dirname(lib_path))
    step_dir = os.path.dirname(src_dir)

    if os.environ.get("BOUKENSHA_DEBUG"):
        print(f"[boukensha] loading from: {step_dir}")

    # Add the step's src directory to sys.path so we can import boukensha
    if src_dir not in sys.path:
        sys.path.insert(0, src_dir)

    # Import and start the REPL
    import boukensha

    if not hasattr(boukensha, "repl"):
        sys.stderr.write(
            f"boukensha: the step at {step_dir}\n"
            f"       does not support the interactive REPL (added in step 08).\n"
            f"       Run its examples directly, e.g.:\n"
            f"         python {step_dir}/examples/*.py\n"
            f"       Or point BOUKENSHA_PATH at step 08 or later.\n"
        )
        sys.exit(1)

    boukensha.repl()


if __name__ == "__main__":
    load_and_start_repl()
